### PR TITLE
Port searchable dropdowns addon

### LINKED ISSUE #163: [New addon] Searchable dropdowns
**Describe the new addon**

This allows you to search all dropdowns in scratch 3.0:
image

**Link to the new addon**
https://github.com/forkphorus/cat-plus
##### Code:

```js
// ==UserScript==
// @name Searchable Dropdowns for Scratch 3
// @version 0.1.2
// @namespace https://github.com/forkphorus/cat-plus
// @match https://scratch.mit.edu/projects/*
// @grant GM_addStyle
// ==/UserScript==

function callback(mutationList) {
  for (const mutation of mutationList) {
    if (mutation.type === 'childList') {
      const newNodes = mutation.addedNodes;
      for (const node of newNodes) {
        if (node.classList && node.classList.contains('blocklyDropdownMenu')) {
          addSearch();
          break;
        }
      }
    }
  }
}

function addSearch() {
  const el = document.createElement('input');
  el.type = 'text';
  el.addEventListener('input', search);
  el.classList.add('u-dropdown-searchbar');
  const container = getDropDownMenu();
  container.insertBefore(el, container.firstChild);
  el.focus();

  for (const child of getItems()) {
    child.dataset.visible = 'true';
  }
}

function search(e) {
  const value = e.target.value.toLowerCase();
  for (const child of getItems()) {
    const text = child.textContent.toLowerCase();
    const contains = text.includes(value);
    child.dataset.visible = contains.toString();
  }
}

let cachedDropDownContentElement = null;
function getDropDownContentElement() {
  if (cachedDropDownContentElement) {
    return cachedDropDownContentElement;
  }
  cachedDropDownContentElement = document.querySelector('.blocklyDropDownContent');
  return cachedDropDownContentElement;
}

function getDropDownMenu() {
  return getDropDownContentElement().querySelector('.blocklyDropdownMenu')
}

function getItems() {
  const el = getDropDownMenu();
  if (el) {
    return Array.from(el.children).filter((child) => child.tagName !== 'INPUT');
  }
  return [];
}

const observer = new MutationObserver(callback);
observer.observe(document.body, {
  childList: true,
  subtree: true,
});

GM_addStyle(`
.u-dropdown-searchbar {
  /* seems to always fill the entire thing without making a horizontal scrollbar */
  width: calc(100% - 20px);
  /* based on styles for the title input */
  color: white;
  background-color: hsla(0, 100%, 100%, 0.25);
  border: 1px dashed hsla(0, 0%, 0%, 0.15);
  padding: .5rem;
  outline: none;
  transition: 0.25s ease-out;
  font-size: 13px;
  font-weight: bold;
  border-radius: 4px;
}
.u-dropdown-searchbar:hover {
  background-color: hsla(0, 100%, 100%, 0.5);
}
.u-dropdown-searchbar:focus {
  background-color: white;
  color: black;
}

[data-visible="false"] {
  display: none;
}

/*.goog-menuitem-highlight {
  background: transparent;
}
.goog-menuitem:hover {
  background: rgba(0, 0, 0, 0.2);
}*/
`);
```

### PR DESCRIPTION
**Pull request by GarboMuffin**
**Resolves**

Part of #70 
Resolves #163 

**Changes**

 - Shows a search bar in any dropdown in the editor
 - Press enter to select the top item
 - Use arrow keys to navigate list, press enter to select
 - Press escape to close

Fc9YOklOxb

(old video: search bar focus being stolen is fixed now, among other things)

**Reason for changes**

#163

**Tests**

Works in Firefox and Chrome on any dropdown.

### CODE REVIEW

**Comment by GarboMuffin** [comment]
I imagine this could be rewritten better using all the hooks that ScratchAddons has.

**Comment by YOYITsM3M8** [comment]
Probably

**Comment by apple502j** [comment]
Currently there are no hooks for scratch-blocks.

**Comment by WorldLanguages** [comment]
> Just remembered that the addon is supposed to export a function

Lol, the quirks of not using `eval`...

> Only observe the part of the DOM that matters

Nice.

> you can't press enter to select the top item

This should be easy, do you want to try and do it? If not, I'll try to do it :P

> hovering over an item in the dropdown steals focus from the text input

Hmm, that one isn't going to be easy, unless @apple502j can think of some magic way?

**Review by WorldLanguages — APPROVED** [review]

**Comment by WorldLanguages** [comment]
👏👏👏👏

**Comment by GarboMuffin** [comment]
I'm happy with the state of this PR right now. All of the major quirks and bugs have been fixed. It works a lot better than I thought it would.

Do some testing and feel free to merge if everything looks good.

**Comment by apple502j** [comment]
For broadcast input, can you make it so that it won't make new broadcast when enter is pressed while focused on empty input?

**Comment by GarboMuffin** [comment]
What should it do instead?

**Comment by apple502j** [comment]
Either ignore or close (without doing anything)

**Comment by apple502j** [comment]
"New Message" is always the first item - perhaps you can check block opcode and the index to ignore? (remember, hidden does not remove it from DOM - index 0 & opcode is broadcast related -> ignore)

File: addons/editor-searchable-dropdowns/userscript.js [hunk 1]
```diff
@@ -92,11 +102,12 @@ export default async function ({ addon, global, console }) {
...
       }
 
+      const selectedBlock = getSelectedBlock();
       const items = getItems();
-      if (event.target.value === "" && blocklyDropDownDiv.dataset.category === "events") {
-        const topItem = items[0];
-        // todo: this won't work in languages other than English
-        if (!topItem.hidden && topItem.innerText === "New message") {
+      if (event.target.value === "" && selectedBlock) {
+        if (selectedBlock.type === "event_broadcast" || selectedBlock.type === "event_whenbroadcastreceived") {
```
Comment (apple502j) [inline addons/editor-searchable-dropdowns/userscript.js:108]: Also broadcast and wait block!

**Pull request merged**